Wifi eswifi, TCP offload listen/accept support

loicpoulain opened the pull request:
Add support for TCP listen/accept

Commit:
wifi: eswifi: Parse async messages

The eswifi controller can generate events via asynchronous messages
which can be polled via the 'MR\r' command. The messages will have a
Start Of Message Asynchronous [SOMA] and End Of Message Asynchronous
[EOMA] delimiters.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
wifi: eswifi: Allow nested locking

In order to prevent potential deadlock in various callbacks such as
accept, recv, etc... Add support for nested eswifi locking, allowing
callee to safely access back to the eswifi methods.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

loicpoulain commented:
Any comments ?

jukkar commented on drivers/wifi/eswifi/eswifi_bus_spi.c:
```diff
@@ -182,15 +182,34 @@ static int eswifi_spi_request(struct eswifi_dev *eswifi, char *cmd, size_t clen,
+		return;
+	}
+
+	if (strncmp(rsp, startstr, strlen(startstr))) {
```
Instead of strlen(), it would be better to do `sizeof(startstr) - 1` as the latter is done at compile time.

loicpoulain replied:
done

jukkar commented on drivers/wifi/eswifi/eswifi_bus_spi.c:
```diff
@@ -182,15 +182,34 @@ static int eswifi_spi_request(struct eswifi_dev *eswifi, char *cmd, size_t clen,
+	}
+
+	/* \r\n[SOMA]...[EOMA]\r\nOK\r\n> */
+	msg_len = ret - strlen(startstr) - strlen(endstr);
```
same here

loicpoulain replied:
done

jukkar commented on drivers/wifi/eswifi/eswifi_bus_spi.c:
```diff
@@ -182,15 +182,34 @@ static int eswifi_spi_request(struct eswifi_dev *eswifi, char *cmd, size_t clen,
+	/* \r\n[SOMA]...[EOMA]\r\nOK\r\n> */
+	msg_len = ret - strlen(startstr) - strlen(endstr);
+	if (msg_len > 0) {
+		eswifi_async_msg(eswifi, rsp + strlen(startstr), msg_len);
```
and here

loicpoulain replied:
done

jukkar commented on drivers/wifi/eswifi/eswifi.h:
```diff
@@ -71,6 +72,8 @@ struct eswifi_dev {
 	char buf[MAX_DATA_SIZE];
 	struct k_mutex mutex;
+	atomic_val_t mutex_owner;
+	unsigned int mutex_depth;
```
Can't you use semaphore instead here as that does counting automatically?

loicpoulain replied:
well, yes, but I there is no limit to the depth, really only need this for counting, maybe a semaphore is overkill here,

jukkar commented on drivers/wifi/eswifi/eswifi_offload.c:
```diff
@@ -125,15 +148,14 @@ static int eswifi_off_bind(struct net_context *context,
 
 	__select_socket(eswifi, socket->index);
 
+	socket->port = (u16_t)sys_be16_to_cpu(net_sin(addr)->sin_port);
```
Casting seems unnecessary here as the value is already 16-bit.

loicpoulain replied:
done

jukkar commented on drivers/wifi/eswifi/eswifi_offload.c:
```diff
@@ -618,6 +663,68 @@ static int eswifi_off_disable_bypass(struct eswifi_dev *eswifi)
+{
+	char msg_tcp_accept[] = "[TCP SVR] Accepted ";
+
+	if (!strncmp(msg, msg_tcp_accept, strlen(msg_tcp_accept))) {
```
sizeof() is better than strlen() here

loicpoulain replied:
done

jukkar commented on drivers/wifi/eswifi/eswifi_offload.c:
```diff
@@ -618,6 +663,68 @@ static int eswifi_off_disable_bypass(struct eswifi_dev *eswifi)
+		int i = 0;
+
+		/* extract client ip/port e.g. 192.168.1.1:8080 */
+		str = msg + strlen(msg_tcp_accept);
```
ditto

loicpoulain replied:
done

jukkar commented on drivers/wifi/eswifi/eswifi_offload.c:
```diff
@@ -618,6 +663,68 @@ static int eswifi_off_disable_bypass(struct eswifi_dev *eswifi)
+		char *str;
+		int i = 0;
+
+		/* extract client ip/port e.g. 192.168.1.1:8080 */
```
Perhaps you could you use `net_ipaddr_parse()` function as that is able parse that kind of strings.

jukkar commented on drivers/wifi/eswifi/eswifi_offload.c:
```diff
@@ -26,6 +26,21 @@ static inline int __select_socket(struct eswifi_dev *eswifi, u8_t idx)
+
+	LOG_DBG("Stopping socket %d", socket->index);
+
+	if (socket->state != ESWIFI_SOCKET_STATE_CONNECTED)
```
{} missing here

jukkar commented on drivers/wifi/eswifi/eswifi_offload.c:
```diff
@@ -302,9 +326,40 @@ static int eswifi_off_accept(struct net_context *context,
+
+	eswifi_unlock(eswifi);
+
+	if (timeout == K_NO_WAIT)
```
{} missing

jukkar commented on drivers/wifi/eswifi/eswifi_offload.c:
```diff
@@ -615,6 +660,69 @@ static int eswifi_off_disable_bypass(struct eswifi_dev *eswifi)
 
+void eswifi_offload_async_msg(struct eswifi_dev *eswifi, char *msg, size_t len)
+{
+	char msg_tcp_accept[] = "[TCP SVR] Accepted ";
```
perhaps "static const" here?

loicpoulain replied:
done

jukkar commented on drivers/wifi/eswifi/eswifi_offload.c:
```diff
@@ -615,6 +660,69 @@ static int eswifi_off_disable_bypass(struct eswifi_dev *eswifi)
+				port = atoi(str);
+				break;
+			}
+			while (*str && (*str != '.') && (*str != ':')) {
```
Please put empty lines after ending block } (in above)

loicpoulain replied:
done

jukkar commented on drivers/wifi/eswifi/eswifi_offload.c:
```diff
@@ -615,6 +660,69 @@ static int eswifi_off_disable_bypass(struct eswifi_dev *eswifi)
+				break;
+			}
+		}
+
```
remove extra empty line

loicpoulain replied:
done

loicpoulain replied:
@jukkar, any other comments?

Commit:
wifi: eswifi: Implement TCP listen/accept

Start es-wifi TCP server on accept.
An asynchronous (spi) message is received on client connection.
Only one connection is supported at a time.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

loicpoulain commented:
rebased
